File: gravitee-apim-cypress/cypress/model/plan.ts
/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an AS IS BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export enum PlanType {
  API = 'API',
  CATALOG = 'CATALOG',
}

export enum MethodType {
  CONNECT = 'CONNECT',
  DELETE = 'DELETE',
  GET = 'GET',
  HEAD = 'HEAD',
  OPTIONS = 'OPTIONS',
  PATCH = 'PATCH',
  POST = 'POST',
  PUT = 'PUT',
  TRACE = 'TRACE',
  OTHER = 'OTHER',
}

export enum OperatorType {
  STARTS_WITH = 'STARTS_WITH',
  EQUALS = 'EQUALS',
}

export interface PathOperator {
  path: string;
  operator: OperatorType;
}

export interface Step {
  name: string;
  policy: string;
  description: string;
  configuration: string;
  enabled: boolean;
}

export enum ConsumerType {
  TAG = 'TAG',
}

export interface Consumer {
  consumerType: ConsumerType;
  consumerId: string;
}

export interface Flow {
  name: string;
  'path-operator': PathOperator;
  pre: Step[];
  post: Step[];
  enabled: boolean;
  methods: MethodType[];
  condition: string;
  consumers: Consumer[];
}

export interface NewPlanEntity {
  id: string;
  name: string;
  description: string;
  validation: PlanValidation;
  security: PlanSecurityType;
  securityDefinition: string;
  type: PlanType;
  status: PlanStatus;
  api: string;
  characteristics: string[];
  tags: string[];
  order: number;
  paths: object;
  flows: Flow[];
  excluded_groups: string[];
  comment_required: boolean;
  comment_message: string;
  selection_rule: string;
  general_conditions: string;
}

export interface Plan {
  id?: string;
  name: string;
  description: string;
  securityDefinition: PlanSecurityType;
  validation?: PlanValidation;
  created_at?: number;
  updated_at?: number;
  status?: PlanStatus;
}

export enum PlanValidation {
  AUTO = 'AUTO',
  MANUAL = 'MANUAL',
}

export enum PlanSecurityType {
  KEY_LESS = 'KEY_LESS',
  API_KEY = 'API_KEY',
  OAUTH2 = 'OAUTH2',
  JWT = 'JWT',
}

export enum PlanStatus {
  STAGING = 'STAGING',
  PUBLISHED = 'PUBLISHED',
  CLOSED = 'CLOSED',
  DEPRECATED = 'DEPRECATED',
}

File: model/src/main/java/io/gravitee/definition/model/Plan.java
/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.definition.model;

import io.gravitee.definition.model.flow.Flow;

import java.io.Serializable;
import java.util.*;

/**
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author GraviteeSource Team
 */
public class Plan implements Serializable {

    private String id;

    private String name;

    private String security;

    private String securityDefinition;

    private Map<String, Path> paths = new HashMap<>();

    private String api;

    private String selectionRule;

    private List<Flow> flows = new ArrayList<>();

    private Set<String> tags;

    private String status;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Map<String, Path> getPaths() {
        return paths;
    }

    public void setPaths(Map<String, Path> paths) {
        this.paths = paths;
    }

    public String getSecurity() {
        return security;
    }

    public void setSecurity(String security) {
        this.security = security;
    }

    public String getSecurityDefinition() {
        return securityDefinition;
    }

    public void setSecurityDefinition(String securityDefinition) {
        this.securityDefinition = securityDefinition;
    }

    public String getApi() {
        return api;
    }

    public void setApi(String api) {
        this.api = api;
    }

    public String getSelectionRule() {
        return selectionRule;
    }

    public void setSelectionRule(String selectionRule) {
        this.selectionRule = selectionRule;
    }

    public List<Flow> getFlows() {
        return flows;
    }

    public void setFlows(List<Flow> flows) {
        this.flows = flows;
    }

    public Set<String> getTags() {
        return tags;
    }

    public void setTags(Set<String> tags) {
        this.tags = tags;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Plan plan = (Plan) o;

        return id.equals(plan.id);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }
}
